### PR TITLE
change pointer free order in amcl to avoid use-after-free bug mentioned in #4068

### PR DESCRIPTION
**Pull request by GoesM**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #4068 |
| Primary OS tested on | Ubuntu22.04 |
| Robotic platform tested on |  gazebo simulation  |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

make `tf_listener_.reset()` before `lase_scan_filter_.reset()`

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Review by SteveMacenski — APPROVED** [review]

File: nav2_amcl/src/amcl_node.cpp [hunk 1]
```diff
@@ -328,6 +328,8 @@ AmclNode::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
   nomotion_update_srv_.reset();
   initial_pose_sub_.reset();
   laser_scan_connection_.disconnect();
+
+  tf_listener_.reset(); // tf_listener_ will access laser_scan_filter, so it should be reset earlier
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:332]: Head the linter

> Error: /__w/navigation2/navigation2/nav2_amcl/src/amcl_node.cpp:332:  At least two spaces is best between code and comments  [whitespace/comments] [2]

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
Also, you need to do the DCO signoff. Click on the failed DCO job for more information

**Comment by GoesM** [comment]
opps, i did `git rebase --signoff` however still face to failed DCO.  So sorry that ,  i'd check it again.

**Comment by SteveMacenski** [comment]
Still both failing (now linter for another reason from change)

**Comment by GoesM** [comment]
It seems that I failed because it need just the only author but regard my local commit as another one author.
Screenshot from 2024-01-26 03-45-35

I'd close this PR and push another again.

sorry again for my unfamiliarity about it.

**Pull request closed without merging**